File: javascript/leetcode/234. Palindrome Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
  if (!head) return true;
  let late = head;
  let early = head.next;
  let prev = null;

  while (early) {
    early = early.next ? early.next.next : null;
    late = late.next;
  }

  let cur = null;
  let mid = late;
  late = null;

  while (mid) {
    cur = mid;
    mid = mid.next;
    cur.next = prev;
    prev = cur;
  }

  while (cur) {
    if (head.val != cur.val) return false;
    head = head.next;
    cur = cur.next;
  }
  return true;
};

File: go/leetcode/234. Palindrome Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
    if head == nil {return true}
    
    late := head
    early := head.Next
    var prev *ListNode 
    
    for early != nil {
        if early.Next != nil {
            early = early.Next.Next
        } else {early = nil}
        late = late.Next
    }
    
    cur := new(ListNode)
    mid := late 
    late = nil
    
    for mid != nil {
        cur = mid
        mid = mid.Next
        cur.Next = prev
        prev = cur
    }
    
    for cur != nil {
        if head.Val != cur.Val {return false}
        head = head.Next
        cur = cur.Next
    }
    
    return true
}   